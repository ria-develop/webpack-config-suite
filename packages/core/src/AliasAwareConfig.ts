import {ResolveAwareConfig} from './ResolveAwareConfig';
import {Configuration} from 'webpack';

export class AliasAwareConfig extends ResolveAwareConfig {
  protected get alias(): Configuration['resolve']['alias'] {
    return {};
  }

  protected get resolve(): Configuration['resolve'] {
    const {alias: currentAlias} = super.resolve;
    const alias = !currentAlias ? this.alias : {...currentAlias, ...this.alias};
    return {...super.resolve, ...(alias ? {alias} : {})};
  }
}
