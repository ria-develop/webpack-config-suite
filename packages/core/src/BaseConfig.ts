import {WebpackCliOptions, What} from './index';

import {Configuration} from 'webpack';

export class BaseConfig {
  //workaround as long as issue opened https://github.com/microsoft/TypeScript/issues/37142
  public get composed(): What[] {
    return [];
  }

  env: unknown;
  argv: WebpackCliOptions;
  config: Configuration;

  constructor(config: Configuration = {}, env: unknown = undefined, argv: WebpackCliOptions = {}) {
    this.env = env;
    this.argv = argv;
    this.config = config;
  }

  protected get isDev(): boolean {
    return this.config.mode === 'development';
  }

  protected get isProd(): boolean {
    return this.config.mode === 'production';
  }

  protected get isSingleOrDevMode(): boolean {
    return !!this.argv.mode || this.isDev;
  }

  protected get isHot(): boolean {
    return this.isSingleOrDevMode && (this.argv.hot || this.argv.hotOnly);
  }

  protected get isLiveReload(): boolean {
    return this.argv.liveReload;
  }
}
