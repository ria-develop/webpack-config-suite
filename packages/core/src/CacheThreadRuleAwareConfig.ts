import {RuleSetAwareConfig} from './RuleSetAwareConfig';
import {RuleSetUseItem} from 'webpack';
import * as os from 'os';

export class CacheThreadRuleAwareConfig extends RuleSetAwareConfig {
  protected get cacheLoader(): RuleSetUseItem {
    return {loader: 'cache-loader'};
  }

  protected get threadLoader(): RuleSetUseItem {
    return {
      loader: 'thread-loader',
      options: {
        poolTimeout: this.isDev ? Infinity : 2000,
        workers: os.cpus().length - 1
      }
    };
  }

  protected get useCache(): boolean {
    return true;
  }

  protected get useThread(): boolean {
    return true;
  }

  protected get use(): RuleSetUseItem[] {
    return super.use.concat([this.useCache && this.cacheLoader, this.useThread && this.threadLoader]).filter(Boolean);
  }
}
