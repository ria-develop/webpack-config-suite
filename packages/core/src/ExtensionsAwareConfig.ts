import {ResolveAwareConfig} from './ResolveAwareConfig';
import {Configuration} from 'webpack';

export class ExtensionsAwareConfig extends ResolveAwareConfig {
  protected get extensions(): Configuration['resolve']['extensions'] {
    return [];
  }

  protected get resolve(): Configuration['resolve'] {
    const {extensions = []} = super.resolve;
    super.resolve.extensions = extensions.concat(this.extensions);
    return super.resolve;
  }
}
