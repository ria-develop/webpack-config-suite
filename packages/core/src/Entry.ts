import {BaseConfig} from './BaseConfig';
import {Configuration, Entry as WebpackEntry} from 'webpack';
import {WebpackCliOptions} from './index';

export class Entry extends BaseConfig {
  protected get sources(): string | [string, ...string[]] {
    return './src';
  }

  protected get entry(): WebpackEntry {
    return this.sources;
  }

  constructor(config: Configuration, env: unknown, argv: WebpackCliOptions) {
    super(config, env, argv);
    const {entry} = this.config;
    this.config.entry =
      entry instanceof Object && this.entry instanceof Object
        ? ({...entry, ...this.entry} as WebpackEntry)
        : this.entry;
  }
}
