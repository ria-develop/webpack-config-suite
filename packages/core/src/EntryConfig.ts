import {BaseConfig} from './BaseConfig';
import {Configuration, Entry} from 'webpack';
import {WebpackCliOptions} from './index';

export class EntryConfig extends BaseConfig {
  protected get sources(): string | [string, ...string[]] {
    return './src';
  }

  protected get entry(): Entry {
    return this.sources;
  }

  constructor(config: Configuration, env: unknown, argv: WebpackCliOptions) {
    super(config, env, argv);
    const {entry} = this.config;
    this.config.entry =
      entry instanceof Object && this.entry instanceof Object ? ({...entry, ...this.entry} as Entry) : this.entry;
  }
}
