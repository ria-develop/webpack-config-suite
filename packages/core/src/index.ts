import {Configuration} from 'webpack';
import {BaseConfig} from './BaseConfig';
import type {OptionsSchema as WebpackCliOptionsSchema} from '../types/webpack-cli';

export * from './BaseConfig';
export * from './AliasAwareConfig';
export * from './EntryConfig';
export * from './ExtensionsAwareConfig';
export * from './JsExtension';
export * from './WebpackSetup';
export * from './ObjectEntryConfig';
export * from './ResolveAwareConfig';
export * from './CacheThreadRuleAwareConfig';
export * from './RuleSetAwareConfig';
export * from './PluginsAwareConfig';
export * from './Utils';

export type {Options as StyleLoaderOptions} from '../types/style-loader';
export type {Options as PostCssLoaderOptions} from '../types/postcss-loader';
export type {PluginOptions as MiniCssExtractPluginOptions} from '../types/mini-css-extract-plugin';
export type {LoaderOptions as ExtractCssLoaderOptions} from '../types/mini-css-extract-loader';
export type {Options as LessLoaderOptions} from '../types/less-loader';
export type {Options as CssLoaderOptions} from '../types/css-loader';
export type {Options as CacheLoaderOptions} from '../types/cache-loader';
export type {Schema as BabelLoaderOptions} from '../types/babel-loader';

export type WebpackCliOptions = {liveReload?: boolean; hot?: boolean; hotOnly?: boolean} & WebpackCliOptionsSchema;

export type What = typeof BaseConfig | ConfigHandler | (typeof BaseConfig | ConfigHandler)[];
export type ConfigHandler = (config: Configuration, env: unknown, argv: WebpackCliOptions) => What[] | void;
