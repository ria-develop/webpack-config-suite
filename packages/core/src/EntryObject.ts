import {Entry} from './Entry';
import {Entry as WebpackEntry} from 'webpack';

export class EntryObject extends Entry {
  protected get dependOn(): string | undefined {
    return undefined;
  }

  protected get name(): string {
    return 'main';
  }

  protected get entry(): WebpackEntry {
    return {[this.name]: this.dependOn ? {import: this.sources, dependOn: this.dependOn} : this.sources};
  }
}
