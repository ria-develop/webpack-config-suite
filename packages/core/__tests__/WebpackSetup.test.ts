import {Configuration} from 'webpack';
import {BaseConfig, WebpackCliOptions, WebpackSetup} from '../src';

class MyBaseConfig extends BaseConfig {
  get name() {
    return 'MyBaseConfig';
  }

  constructor(config: Configuration, env: unknown, argv: WebpackCliOptions) {
    super(config, env, argv);
    if (!config.name) {
      config.name = this.name;
    } else {
      config.name = `${config.name} ${this.name}`;
    }
  }
}

class MyConfig extends MyBaseConfig {
  get name() {
    return 'MyConfig';
  }

  get composed() {
    return [MyAnotherConfig];
  }
}

class MyAnotherConfig extends MyBaseConfig {
  get composed() {
    return undefined;
  }

  get name() {
    return 'MyAnotherConfig';
  }
}

describe.each([
  ['development', undefined, {}],
  ['production', undefined, {mode: 'production'}]
])(
  'given mode: "%s", env: "%j", argv: "%j"',
  (mode: Configuration['mode'], env: unknown, argv: WebpackCliOptions): void => {
    it('should match snapshot', () => {
      expect(
        WebpackSetup.newSetup()
          .use(MyConfig)
          .use(MyConfig) //should process it only once
          .use((config) => {
            config.name = `${config.name} MyConfigModifierFunction`;
            return [
              (config) => {
                config.name = `${config.name} MyConfigModifierNestedFunction`;
              }
            ];
          })
          .newFactory()(env, argv)
      ).toMatchSnapshot();
    });
  }
);

describe('given empty setup', () => {
  it('should match snapshot', () => {
    expect(WebpackSetup.newSetup().newFactory()(undefined, {})).toMatchSnapshot();
  });
});

describe('given multiple configs for use', () => {
  it('should match snapshot', () => {
    expect(WebpackSetup.newSetup().use([MyConfig, MyConfig]).newFactory()(undefined, {})).toMatchSnapshot();
  });
});
