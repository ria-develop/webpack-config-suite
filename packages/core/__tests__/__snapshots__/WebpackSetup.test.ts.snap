// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given empty setup should match snapshot 1`] = `
Array [
  Object {
    "mode": "production",
  },
  Object {
    "mode": "development",
  },
]
`;

exports[`given mode: "development", env: "undefined", argv: "{}" should match snapshot 1`] = `
Array [
  Object {
    "mode": "production",
    "name": "MyConfig MyAnotherConfig MyConfigModifierFunction MyConfigModifierNestedFunction",
  },
  Object {
    "mode": "development",
    "name": "MyConfig MyAnotherConfig MyConfigModifierFunction MyConfigModifierNestedFunction",
  },
]
`;

exports[`given mode: "production", env: "undefined", argv: "{"mode":"production"}" should match snapshot 1`] = `
Array [
  Object {
    "mode": "production",
    "name": "MyConfig MyAnotherConfig MyConfigModifierFunction MyConfigModifierNestedFunction",
  },
]
`;

exports[`given multiple configs for use should match snapshot 1`] = `
Array [
  Object {
    "mode": "production",
    "name": "MyConfig MyAnotherConfig",
  },
  Object {
    "mode": "development",
    "name": "MyConfig MyAnotherConfig",
  },
]
`;
